### PR TITLE
Adding validation for illegal use of aggregation functions

### PR DESCRIPTION
**Pull request by fulghum**
GMS currently panics if an aggregation function that relies on the implicit, all-rows grouping is mixed aggregation functions that rely on a window. This PR adds validation to detect these invalid queries and return the exact same error message as MySQL. 

Fixes: https://github.com/dolthub/dolt/issues/4739

### CODE REVIEW

**Review by max-hoffman — APPROVED** [review]
LGTM, i generally think the MySQL error message is confusing and some form of "window and group by not supported in the same clause" would make more sense? The ergonomics of checking this condition here and not earlier in `parse.go` makes this more difficult than it maybe has to be, also. Food for thought whenever we get around to reworking semantic validation.

File: sql/analyzer/validation_rules.go [hunk 1]
```diff
@@ -683,39 +683,100 @@ func validateReadOnlyTransaction(ctx *sql.Context, a *Analyzer, n sql.Node, scop
...
-					invalidExpr = e
-				}
-				return invalidExpr == nil
-			})
-		}
-		return invalidExpr == nil
-	}
+	var validationErr error
 	transform.Inspect(n, func(n sql.Node) bool {
 		if gb, ok := n.(*plan.GroupBy); ok {
```
[the remark this answers is not in the document]
Comment (fulghum) [reply sql/analyzer/validation_rules.go]: I "switched" it for ya 😉  Adds a line or two, but I think it's slightly easier to read this way.

File: sql/errors.go [hunk 2]
```diff
@@ -444,6 +444,12 @@ var (
 	// ErrTableFunctionNotFound is thrown when a table function is not found
 	ErrTableFunctionNotFound = errors.NewKind("table function: '%s' not found")
 
+	// ErrNonAggregatedColumnWithoutGroupBy is thrown when an aggregate function is used with the implicit, all-rows
+	// grouping and another projected expression contains a non-aggregated column.
+	// MySQL error code: 1140, SQL state: 42000
+	ErrNonAggregatedColumnWithoutGroupBy = errors.NewKind("in aggregated query without GROUP BY, expression #%d of SELECT list contains nonaggregated column '%s'; " +
```
Comment (zachmu) [inline sql/errors.go:450]: Do we have an error code mapping for this?

**Review by zachmu — COMMENTED** [review]
I think in general I'm for copying MySQL error messages verbatim where the error semantics are the same, but I don't have strong feelings about this and we aren't consistent in applying this principle

File: sql/errors.go (hunk 2 above)
Comment (fulghum) [reply sql/errors.go:450]: Good call; added in https://github.com/dolthub/go-mysql-server/pull/1400.

File: sql/analyzer/validation_rules.go [hunk 3]
```diff
@@ -683,39 +683,102 @@ func validateReadOnlyTransaction(ctx *sql.Context, a *Analyzer, n sql.Node, scop
...
 // aggregation expressions end up in the wrong place.
 func validateAggregations(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	var invalidExpr sql.Expression
-	checkExpressions := func(exprs []sql.Expression) bool {
-		for _, e := range exprs {
-			sql.Inspect(e, func(ie sql.Expression) bool {
-				if _, ok := ie.(sql.Aggregation); ok {
-					invalidExpr = e
+	var validationErr error
+	transform.Inspect(n, func(n sql.Node) bool {
```
[the remark this answers is not in the document]
Comment (fulghum) [reply sql/analyzer/validation_rules.go:697]: Good question... this won't descend into opaque nodes, but the `finalizeSubqueries` rule will run the validation rules on derived tables and catch errors and `finalizeUnions` will do the same for opaque union nodes. Recursive CTE opaque node handling seems to work correctly as well. Good call to add tests for those; they are in: PR https://github.com/dolthub/go-mysql-server/pull/1400.

**Pull request merged**